File: vpbuf/examples/poker/rel1/python/draw.py
#!/usr/bin/python3
# Software released under the MIT license (see project root for license file)

from pyamf import amf3
from vp_poker.persist import *
from vp_poker.vp_poker import *
from vp_poker.tools import *

import argparse

# ------------------------------------------------------------------------------

def main():
    # process args

    p = argparse.ArgumentParser()
    p.add_argument("deck_name")
    p.add_argument("count", nargs='?', default=1)

    args = p.parse_args()

    print('draw', args.count)

    # load the specified deck

    deck = load_deck(args.deck_name)

    if (deck is None):
        exit(1)

    if (len(deck.cards) < int(args.count)):
        print('too few cards left')
    else:
        # modify the deck by removing cards

        print('(python) draw :', end='')
        for i in range(0, int(args.count)):
            c = deck.cards.pop()
            print(' ' + c.get_name(), end='')
        print()

        # write the modified deck back out

        save_deck(args.deck_name, deck)

if __name__ == '__main__':
    main()

# ------------------------------------------------------------------------------

File: vpbuf/examples/poker/rel1/cpp/src/draw.cc
// Software released under the MIT license (see project root for license file)

#include <cstddef>
#include <iostream>
#include <fstream>
#include <vector>
#include <cassert>
#include <typeindex> // needed for class polymorphism

using namespace std;

typedef char byte;
typedef vector<byte> ByteVec;

// buffer management helper classes and functions
#include "buffer_context.cc"

// include classes to be serialized
#include "persist.cc"

// include output from vpc code generator
#include "vp_poker.cc"
using namespace vp_poker;

// -----------------------------------------------------------------------------

int main(int argc, char* argv[])
{
   string deck_name;
   int count = 1;

   if (argc == 1) {
      cout << "usage: draw <deck_name> [<card_count>]\n";
   } else {
      deck_name = argv[1];

      if (argc == 3)
         count = stoi(argv[2]);

      cout <<"draw "<< count << endl;

      // load the specified deck

      Deck *deck = load_deck(deck_name);

      // modify the deck by removing cards

      if (deck->cards.size() < count) {
         cout << "too few cards remaining\n";
      } else {
         cout <<"(cpp) draw:";
         for (auto i = 0; i < count; i++) {
            Card *c =  deck->cards.back();
            cout <<" "<< c->get_name();
            deck->cards.pop_back();
         }
         cout <<endl;

         // write the modified deck back out

         save_deck(deck_name, *deck);

         cout <<"cards left "<< deck->cards.size() <<endl;
      }
   }
}

// -----------------------------------------------------------------------------